src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -1733,4 +1806,109 @@ public void makeIndexQueryable(Index index, Index.Status status)
+            }
+
+            Map<String, Index.Status> oldStatus = peerIndexStatus.put(endpoint, indexStatus);
+            Map<String, Index.Status> updated = updatedIndexStatuses(oldStatus, indexStatus);
+            Set<String> removed = removedIndexStatuses(oldStatus, indexStatus);
+            if (!updated.isEmpty() || !removed.isEmpty())
+                logger.debug("Received index status for peer {}:\n    Updated: {}\n    Removed: {}",
+                             endpoint, updated, removed);
+        }
+        catch (Throwable e)
+        {
+            logger.warn("Unable to parse index status: {}", e.getMessage());
+        }
+    }
+
+    /**
+     * Returns the names of indexes that are present in oldStatus but absent in newStatus.
+     */
+    private static @Nonnull Set<String> removedIndexStatuses(@Nullable Map<String, Index.Status> oldStatus, @Nonnull Map<String, Index.Status> newStatus)
+    {
+        if (oldStatus == null)
+            return Collections.emptySet();
+        Set<String> result = new HashSet<>(oldStatus.keySet());
+        result.removeAll(newStatus.keySet());
+        return result;
+    }
+
+    /**
+     * Returns a new map containing only the entries from newStatus that differ from corresponding entries in oldStatus.
+     */
+    private static @Nonnull Map<String, Index.Status> updatedIndexStatuses(
+    @Nullable Map<String, Index.Status> oldStatus,
+    @Nonnull Map<String, Index.Status> newStatus)
+    {
+        Map<String, Index.Status> delta = new HashMap<>();
+        for (Map.Entry<String, Index.Status> e : newStatus.entrySet())
+        {
+            if (oldStatus == null || e.getValue() != oldStatus.get(e.getKey()))
+                delta.put(e.getKey(), e.getValue());
+        }
+        return delta;
+    }
+
+    private synchronized static void propagateLocalIndexStatus(String keyspace, String index, Index.Status status)
+    {
+        try
+        {
+            Map<String, Index.Status> states = peerIndexStatus.computeIfAbsent(FBUtilities.getBroadcastAddressAndPort(),
+                                                                               k -> new ConcurrentHashMap<>());
+            String keyspaceIndex = identifier(keyspace, index);
+
+            if (status == Index.Status.DROPPED)
+                states.remove(keyspaceIndex);
+            else
+                states.put(keyspaceIndex, status);
+
+            // Don't try and propagate if the gossiper isn't enabled. This is primarily for tests where the
+            // Gossiper has not been started. If we attempt to propagate when not started an exception is
+            // logged and this causes a number of dtests to fail.
+            if (Gossiper.instance.isEnabled())
+            {
+                String newStatus = JSONValue.toJSONString(states.entrySet().stream()
+                                                                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().toString())));
```

Comment:
You might not want to create the mapper anew every time, but `new ObjectMapper().writeValueAsString(states);` would do this by default, without the stream stuff.